### PR TITLE
feat(api): add selectors for easier selection of columns

### PR DESCRIPTION
**Pull request by cpcloud**
Convenient columnn selectors, shamelessly cribbed from https://dplyr.tidyverse.org/reference/select.html.

### CODE REVIEW

**Comment by kszucs** [comment]
Nice! Are you still working on this?

**Comment by cpcloud** [comment]
@kszucs Just finished up the first iteration of the docs. It's ready for review now.

**Comment by jreback** [comment]
pretty neat @cpcloud

File: ibis/expr/selectors.py [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+from typing import Callable, Iterable, Sequence
+
+from public import public
+
+import ibis.expr.datatypes as dt
+import ibis.expr.types as ir
+from ibis import util
+
+
+class Selector:
```
Comment (kszucs) [inline ibis/expr/selectors.py:56]: Could we subclass deferred here? I mean a deferred base to not provide all of the overloads. 

Mostly just asking, shouldn't block this PR.

**Review by kszucs — APPROVED** [review]
Nice API addition!

File: ibis/expr/selectors.py (hunk 1 above)
Comment (cpcloud) [reply ibis/expr/selectors.py:56]: Maybe, but I didn't think much about what the semantics of the other overloads would be. It did cross my mind that maybe there's some shared code between this and `Deferred` but I didn't want to do a bunch of refactoring here.

**Comment by krzysztof-kwitt** [comment]
I wonder if this API will be easy to use in real world., so I'm looking at the Ibis code that used column groups. Is it possible to rewrite this piece of code to use this new API? 
https://github.com/ibis-project/ibis/blob/318179add952a3b2fc80a1b6bf82191dc84b5157/docs/backends/app/backend_info_app.py#L155-L158

**Comment by cpcloud** [comment]
> I wonder if this API will be easy to use in real world., so I'm looking at the Ibis code that used column groups. Is it possible to rewrite this piece of code to use this new API?
> 
> https://github.com/ibis-project/ibis/blob/318179add952a3b2fc80a1b6bf82191dc84b5157/docs/backends/app/backend_info_app.py#L155-L158

I don't think so, because that bit of code isn't filtering the columns by name or property in any way, it's looking at every column in the table.

**Comment by mik-laj** [comment]
Can we use this API to calculate the number of boolean columns and count the number of True values?

**Comment by cpcloud** [comment]
You could do something like:

```python
sum(map(t.__getitem__, t.select(s.of_type("bool")).columns))
```

**Comment by cpcloud** [comment]
I wonder if it makes sense to define `TableExpr.__iter__` to iterate over the column objects, which which make the computation much shorter:

```python
sum(t.select(s.of_type("bool")))
```

**Comment by cpcloud** [comment]
I know some systems are starting to allow concise syntax for doing whole-table operations. Maybe we can revisit adding aggregations and some kind of `apply expression to every column` method, so you can do things like:

```python
t.select(s.of_type("bool")).cast("int").sum()  # single row table, cast applies to every column
```

**Comment by cpcloud** [comment]
@mik-laj You might find https://dplyr.tidyverse.org/articles/colwise.html interesting. Those APIs are perhaps a good place to start experimenting with APIs for summarizing things across a subset of the columns in a table.

I can imagine something like:

```python
# group by a, b and count all string columns excluding a and b
t.group_by([_.a, _.b]).summarize(s.across(s.of_type("string") & !s.contains(("a", "b")), [_.count()]))
```

**Comment by jcmkk3** [comment]
> @mik-laj You might find https://dplyr.tidyverse.org/articles/colwise.html interesting. Those examples are perhaps a good place to start experimenting with APIs for summarizing things across a subset of the columns in a table.

It might also be worth taking a look at the [polars API](https://pola-rs.github.io/polars/py-polars/html/reference/expressions/api/polars.col.html) for this. Basically, the `pl.col` selector allows selecting multiple columns in a few different ways, then when you perform methods on that selection it will apply to all columns that were selected. dplyr has to approach the problem a bit differently since the data transformations are functions instead of methods.

```python
# group by a, b and count all string columns excluding a and b
t.groupby(["a", "b"]).agg(pl.col(pl.Utf8).exclude(["a", "b"]).count())
```

**Comment by cpcloud** [comment]
Neat, didn't know `pl.col` did all of that!

**Comment by cpcloud** [comment]
Can you comment more on why the method/function distinction matters for anything other than than the API design (as opposed to being an implementation detail)?

I would expect that the fundamentals of the problem are unrelated to how a thing is called, e.g., a function vs a method on an object.

**Comment by jcmkk3** [comment]
> Can you comment more on why the method/function distinction matters for anything other than than the API design (as opposed to being an implementation detail)?

I think that you're generally right in that it is a design constraint and also differences in style between the languages. R *can* do object oriented programming, but it is not the norm (especially in the tidyverse ecosystem).

The dplyr package doesn't own the compute functions like Ibis does so it makes a lot of sense for them to implement the `across` helper as a higher order function.

I personally think that something like the polars approach is more natural in python (minus the `Utf8` type that should probably be `Str`). It could still be a function like `across` that originally selects multiple columns, but the transformations/computations would be applied through method chaining just like working with a single column.

**Comment by ksuarez1423** [comment]
While `across` seems to fulfill the need quite nicely, I agree that the Polars approach feels more natural in Python. When I went to try the selector with an aggregate, my instinct was to call `t.select(s.numeric()).min()`, because of its similarity to `t["col"].min()`. From my perspective, `pl.col()` matches that experience well enough for it to feel natural.

Table-wide operations also make sense from a user perspective, because it means if I have a fully numeric table, I could do `t.sum()` or even `t.all().sum()` if that would make code easier. Then, that makes `t.select(s.numeric()).sum()` a very easy counterpart -- I'd be able to fairly easily think, "oh, `t.select()` gets me a subset that's a table, so all the table-wide things work!"

**Comment by cpcloud** [comment]
`pl.col` has a bit too much going on for my taste but I do agree that method chaining to build up the selectors is the way to go.

**Comment by jcmkk3** [comment]
> `pl.col` has a bit too much going on for my taste but I do agree that method chaining to build up the selectors is the way to go.

I agree with this as well. I think that starting with a specific function for performing an operation on multiple columns would stand out more when scanning the code.

**Pull request merged**